core/primitives/src/account.rs

```diff
@@ -85,80 +60,140 @@ impl TryFrom<account_proto::Account> for Account {
+                Some(access_key_proto::AccessKey_oneof_permission::full_access(_)) => {
+                    AccessKeyPermission::FullAccess
+                }
+                Some(access_key_proto::AccessKey_oneof_permission::function_call(fc)) => {
+                    AccessKeyPermission::FunctionCall(FunctionCallPermission::try_from(fc)?)
+                }
+                None => return Err("No such permission".into()),
+            },
         })
     }
 }
 
 impl From<AccessKey> for access_key_proto::AccessKey {
     fn from(access_key: AccessKey) -> Self {
+        let permission = match access_key.permission {
+            AccessKeyPermission::FullAccess => {
+                access_key_proto::AccessKey_oneof_permission::full_access(
+                    access_key_proto::FullAccessPermission::default(),
+                )
+            }
+            AccessKeyPermission::FunctionCall(fc) => {
+                access_key_proto::AccessKey_oneof_permission::function_call(fc.into())
+            }
+        };
         access_key_proto::AccessKey {
-            amount: SingularPtrField::some(access_key.amount.into()),
-            balance_owner: SingularPtrField::from_option(access_key.balance_owner.map(|v| {
-                let mut res = StringValue::new();
-                res.set_value(v);
-                res
-            })),
-            contract_id: SingularPtrField::from_option(access_key.contract_id.map(|v| {
+            nonce: access_key.nonce,
+            permission: Some(permission),
+
+            cached_size: Default::default(),
+            unknown_fields: Default::default(),
+        }
+    }
+}
+
+/// Grants limited permission to issue transactions with a single function call action.
+/// Those function calls can't have attached balance.
+/// The permission can limit the allowed balance to be spent on the prepaid gas.
+/// It also restrict the account ID of the receiver for this function call.
+/// It also can restrict the method name for the allowed function calls.
+#[derive(Serialize, Deserialize, PartialEq, Eq, Hash, Clone, Debug)]
+pub struct FunctionCallPermission {
+    /// Allowance is a balance limit to use by this access key to pay for function call gas and
+    /// transaction fees. When this access key is used, both account balance and the allowance is
+    /// decreased by the same value.
+    /// `None` means unlimited allowance.
+    /// NOTE: To change or increase the allowance, the old access key needs to be deleted and a new
+    /// access key should be created.
+    #[serde(with = "option_u128_dec_format")]
+    pub allowance: Option<Balance>,
+
+    /// The access key only allows transactions with the given receiver's account id.
+    pub receiver_id: AccountId,
+
+    /// `Some` means the access key only allows transactions with the function call of the given
+    /// method name.
+    /// `None` means any method name can be used.
+    pub method_name: Option<String>,
```

Comment:
Why do we not allow restricting to multiple methods? It seems inflexible

Reply:
Linear check would be sub-optimal and might introduce transaction verification in non-constant time. It's probably acceptable to support reasonable number of method_names, e.g. up to N or up to LEN total number of characters. Since we still need to deserialize AccessKey each time we're trying to validate a transaction.

It follows the NEP right now. Restricting method_name is done to support `proxy` function, which is enough for most use-cases.

Reply:
As discussed offline we're switching to Vec<String>